### PR TITLE
[Concept]: Basics

### PR DESCRIPTION
**Pull request by glaxxie**
- Basics for PowerShell : short history, introductions
- Concept exercise : classic lasagna exercise
- Extra: eternal links for related concepts mentioned

### CODE REVIEW

File: concepts/basics/about.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
+# About
+
+PowerShell is a dynamically implicit typed (also support strongly typed variables) scripting language from Microsoft, mostly focus on task automation and confifuration management.
```
Comment (BethanyG) [inline concepts/basics/about.md]:
```suggestion
PowerShell is a dynamic and implicitly typed (_but also supporting [strongly typed][strong typed] variables_) scripting language from Microsoft, mostly focused on task automation and configuration management.
```

File: concepts/basics/.meta/config.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+    "blurb": "PowerShell is a task automation and configuration management program from Microsoft, consisting of a command-line shell and the associated scripting language.",
```
Comment (IsaacG) [inline concepts/basics/.meta/config.json]: I've never heard a shell described this way before

File: concepts/basics/about.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+PowerShell supports imperative, functional, object-oriented programming, and even pipeline style.
```
Comment (BethanyG) [inline concepts/basics/about.md]:
```suggestion
PowerShell supports imperative, functional, object-oriented, and even pipeline style programming.
```

File: concepts/basics/about.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+PowerShell was built on the .NET platform and first released in 2006.
+It has been included by default in all modern Windows OS, and can be installed on Mac and Linux.
```
Comment (BethanyG) [inline concepts/basics/about.md]:
```suggestion
It has been included by default in all modern Windows OS, and can also be installed on Mac and Linux.
```

File: concepts/basics/about.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+1
+
+# String of letters can not be converted to double type
+> [double]$staticTwo = "hello"
+MetadataError: Cannot convert value "hello" to type "System.Double"
+```
+
+### Constants
+
+Constants in PowerShell are read-only variables that can't be **removed** after defined.
```
Comment (IsaacG) [inline concepts/basics/about.md]: This implies they can have their value changed, just not removed.

File: concepts/basics/about.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+PowerShell supports imperative, functional, object-oriented programming, and even pipeline style.
+
+PowerShell was built on the .NET platform and first released in 2006.
+It has been included by default in all modern Windows OS, and can be installed on Mac and Linux.
+
+## Variables
+
+There are two common ways to define a variable in PowerShell.
+
+Varible name in PowerShell must start with `$` and using the assignment operator `=` to bind the variable name to a value.
```
Comment (BethanyG) [inline concepts/basics/about.md]:
```suggestion
Variable names in PowerShell must start with a `$`, and use the assignment operator `=` to bind the variable name to a value.
```

File: concepts/basics/about.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+> $MyConstant
+100
+```
+
+For a more detailed and advanced way to create a new variable with various options, please read [New-Variable][New-Variable].
+
+## Functions
+
+Functions in PowerShell are defined using the `function` keyword.
+Each function can have zero or more formal parameters in `()` parenthesis, followed by its own scope start  with `{` and end with `}`.
```
Comment (IsaacG) [inline concepts/basics/about.md]:
```suggestion
Each function can have zero or more formal parameters in `()` parenthesis, followed by its own scope start with `{` and end with `}`. 
```

File: concepts/basics/about.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+> $intVar = 1
+> $doubleVar = 2.0
+> $stringVar = "Hello"
+
+# Checking a variable type using the GetType() method
+> Write-Host $doubleVar.GetType()
+System.Double
+```
+
+You can also explicitly specify a variable's type if you want to enforce a specific type.
```
Comment (BethanyG) [inline concepts/basics/about.md]:
```suggestion
You can also specify a variable's type explicitly, if you want to enforce/require the value be of a certain type.
```

File: concepts/basics/about.md [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+Constants in PowerShell are read-only variables that can't be **removed** after defined.
+This lead us to the second way to define variables using the cmdlet `New-Variable`.
+
+```powershell
+> New-Variable -Name "MyConstant" -Value 100 -Option Constant
+> $MyConstant
+100
+```
+
+For a more detailed and advanced way to create a new variable with various options, please read [New-Variable][New-Variable].
```
Comment (BethanyG) [inline concepts/basics/about.md]:
```suggestion
For more detailed and advanced ways to create a new variable, please read [New-Variable][New-Variable].
```

File: exercises/concepts/jeffrey-juicy-lasagna/.meta/design.md [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+- Detailed documentation comments.
+- Memory and performance characteristics.
+- Automatic variables.
+- Pipe operation.
+
+## Concepts
+
+`basics`:
+
+- function definition & `return` keyword
```
Comment (IsaacG) [inline exercises/concepts/jeffrey-juicy-lasagna/.meta/design.md]:
```suggestion
- function definition and `return` keyword
```

File: concepts/basics/about.md (hunk 7 above)
Comment (BethanyG) [inline concepts/basics/about.md]:
```suggestion
Each function can have zero or more formal parameters in `()` parenthesis, followed by the function scope starting  with `{` and ending with `}`. 
```

File: exercises/concepts/jeffrey-juicy-lasagna/.meta/design.md [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+- integer, double, string
+- mathematical operators
+- single line comments
+- updating variable values
+- using functions
+- variables, variable assignment, & variable naming
```
Comment (IsaacG) [inline exercises/concepts/jeffrey-juicy-lasagna/.meta/design.md]:
```suggestion
- variables, variable assignment and variable naming
```

Use of the Oxford comma should be consistent

**Review by IsaacG — COMMENTED** [review]
How much English nit picking and wording suggestions do you want?

File: concepts/basics/about.md [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+Functions in PowerShell are defined using the `function` keyword.
+Each function can have zero or more formal parameters in `()` parenthesis, followed by its own scope start  with `{` and end with `}`. 
+
+```powershell
+function Greeting {
+    return "Hello!"
+}
+```
+
+Functions in PowerShell can return values explicitly via the `return` keyword, after being called using the syntax `FuncName $Arg`.
```
Comment (BethanyG) [inline concepts/basics/about.md:64]: Is there an implicit return when `return` is omitted?

File: concepts/basics/about.md [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    $result = $NumOne + $NumTwo
+    return $result
+
+> AddTwoNumbers 10 20
+30
+```
+
+## Comments
+
+Single comment in PowerShell start with a `#`, and end at line termination.
```
Comment (BethanyG) [inline concepts/basics/about.md]:
```suggestion
Single comments in PowerShell start with a `#`, and end at line termination.
```

File: concepts/basics/about.md [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+Comment block in PowerShell start with `<#` and end with `#>`.
```
Comment (BethanyG) [inline concepts/basics/about.md]:
```suggestion
Comment blocks in PowerShell start with `<#` and end with `#>`.
```

File: concepts/basics/about.md [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+```powershell
+function Main() {
+    #This is a comment
+}
+```
```
Comment (BethanyG) [inline concepts/basics/about.md]: Feels like there should also be a mulit-line example here, to demonstrate what that looks like.

Comment (BethanyG) [inline concepts/basics/about.md]: Feels like there should also be a mulit-line example here, to demonstrate what that looks like.

File: concepts/basics/introduction.md
Comment (BethanyG) [inline concepts/basics/introduction.md:1]: Same suggestions here as made in the `about.md`

File: config.json [hunk 16]
```diff
@@ -34,7 +34,16 @@
     ]
   },
   "exercises": {
-    "concept": [],
+    "concept": [
+        {
+        "slug": "jeffrey-juicy-lasagna",
+        "name": "Jeffrey Juicy Lasagna",
```
Comment (BethanyG) [inline config.json]:
```suggestion
        "name": "Jeffreys Juicy Lasagna",
```

File: config.json [hunk 17]
```diff
@@ -34,7 +34,16 @@
     ]
   },
   "exercises": {
-    "concept": [],
+    "concept": [
+        {
+        "slug": "jeffrey-juicy-lasagna",
```
Comment (BethanyG) [inline config.json]:
```suggestion
        "slug": "jeffreys-juicy-lasagna",
```

File: exercises/concepts/jeffrey-juicy-lasagna/.docs/hints.md [hunk 18]
```diff
@@ -0,0 +1,42 @@
+# Hints
+
+## General
+- A variable [type][types] are based on the value of that variables: *10* as `Int` , *'Hello'* as `String`, or *2.0* as `double`.
```
Comment (BethanyG) [inline exercises/concepts/jeffrey-juicy-lasagna/.docs/hints.md]:
```suggestion
- A variable's [type][types] is based on the value of that variable: *10* as `Int` , *'Hello'* as `String`, or *2.0* as `double`.
```

File: exercises/concepts/jeffrey-juicy-lasagna/.docs/hints.md [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+
+## 1. Define constant expected bake time in minutes
+
+- You need to define a constant variable, and assign it an integer value.
```
Comment (BethanyG) [inline exercises/concepts/jeffrey-juicy-lasagna/.docs/hints.md]:
```suggestion
- You need to define a constant variable and assign it an integer value.
```

File: exercises/concepts/jeffrey-juicy-lasagna/.docs/hints.md [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+- Constant can be defined through [New-Variable cmdlet][new variable].
```
Comment (BethanyG) [inline exercises/concepts/jeffrey-juicy-lasagna/.docs/hints.md]:
```suggestion
- Constants can be defined through the [New-Variable cmdlet][new variable].
```

File: concepts/basics/about.md (hunk 5 above)
Comment (BethanyG) [inline concepts/basics/about.md]:
```suggestion
Constants in PowerShell are read-only variables that can't be **removed** after definition.
```

File: concepts/basics/about.md [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+
+# String of letters can not be converted to double type
+> [double]$staticTwo = "hello"
+MetadataError: Cannot convert value "hello" to type "System.Double"
+```
+
+### Constants
+
+Constants in PowerShell are read-only variables that can't be **removed** after defined.
+This lead us to the second way to define variables using the cmdlet `New-Variable`.
```
Comment (BethanyG) [inline concepts/basics/about.md]:
```suggestion
This leads us to the second way of defining variables: using the cmdlet `New-Variable`.
```

File: exercises/concepts/jeffrey-juicy-lasagna/.meta/config.json [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+      "exemplar": [
+        ".meta/lasagna.example.ps1"
+      ]
+    },
+    "forked_from": [
+      "csharp/lucians-luscious-lasagna",
+      "python/guidos-gorgeous-lasagna"
+    ],
+    "icon": "lasagna",
+    "blurb": "Learn the basics of Python by making lasagna."
```
Comment (BethanyG) [inline exercises/concepts/jeffrey-juicy-lasagna/.meta/config.json]:
```suggestion
    "blurb": "Learn the basics of Powershell by making lasagna."
```

File: exercises/concepts/jeffrey-juicy-lasagna/.meta/design.md [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+The goal of this exercise is to teach the student some of the basics of programming with Powershell.
+
+## Learning objectives
+
+- Know what a variable is and how to define it in many ways.
+- Know how to update or reassign variable.
+- Know that variable can have type conversion.
+- Know what a `function` is, its basic interface and how to define one via `function`.
+- Know how to add one or more `parameters` to a function definition.
+- Know that `parameters` definitions can have type annotations.
```
Comment (BethanyG) [inline exercises/concepts/jeffrey-juicy-lasagna/.meta/design.md]: I didn't see either of these in the exercise instructions, but maybe in the links?

**Review by BethanyG — APPROVED** [review]
Excellent work!  I've left suggestions/comments mostly on the `about.md` file, with the same comments applying to where `about.md` overlaps with the `introduction.md` files.  Most of these are _suggestions_ - so if they don't sound or feel right to you, please feel free to ignore them.  Except for the change from Python --> Powershell in the `config.json`.  😉  That you probably want to make.  😄

File: concepts/basics/about.md (hunk 5 above)
Comment (glaxxie) [reply concepts/basics/about.md]: I thought the read-only mention already addressed that? But yeah I could change the wording here

File: concepts/basics/.meta/config.json (hunk 2 above)
Comment (glaxxie) [reply concepts/basics/.meta/config.json]: Oh I lift this one straight from Wikipedia for a quick blurb 😆

File: exercises/concepts/jeffrey-juicy-lasagna/.meta/design.md (hunk 23 above)
Comment (glaxxie) [reply exercises/concepts/jeffrey-juicy-lasagna/.meta/design.md]: Oh yeah, that 2nd part should be moved to the advanced function concept.

**Review by ErikSchierboom — APPROVED** [review]
Really nice!

**Pull request merged**